dependency first inflator for Xaml sourcegen

StephaneDelcroix opened the pull request.

StephaneDelcroix committed:
incremental, dependency first, sourcegen inflator

benefits:
- creates (a lot) of smaller methods, easier, faster to jit
- provides a solutions for forward references
- only creates objects that ends up being used
- better code gen for extensions and converters
- possible to hotrelaod using c# hotreload

simonrozsival commented on src/Controls/src/SourceGen/LocalVariable.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Microsoft.Maui.Controls.SourceGen;
+
+interface ILocalVariable
```
I only see one implementation of this interface (or did I miss it?). What's the plan with `ILocalVariable`?

simonrozsival commented on src/Controls/src/SourceGen/InitializeComponentCodeWriter.cs:
```diff
@@ -87,21 +87,35 @@ PrePost newblock() =>
 				xamlItem.Root!.XmlType.TryResolveTypeSymbol(null, compilation, xmlnsCache, out var baseType);
 				var sgcontext = new SourceGenContext(codeWriter, compilation, sourceProductionContext, xmlnsCache, typeCache, rootType!, baseType, xamlItem.ProjectItem);
-				using (newblock())
+				if (xamlItem.ProjectItem.TreeOrder)
```
I'm a little confused. Why do we want to keep the previous way of generating code around? Don't we want to fully replace it with the new codegen? Is this just temporary until we fix all the existing unit tests?

StephaneDelcroix committed:
staticresource

StephaneDelcroix committed:
setter

simonrozsival committed:
AppThemeBindingExtension

StephaneDelcroix committed:
[XSG] various perf enhancement

perf enhancement for BindingExtension, StaticResourceExtension,
BindablePropertyConverter, TyptTypeConverter

also prepare the code for what's next

StephaneDelcroix committed:
[XSG] Benchmark XAML inflators

StephaneDelcroix committed:
VSM

StephaneDelcroix committed:
knownmarkups can gen more than DirectValue

StephaneDelcroix committed:
benchmark the time spend in the generator

StephaneDelcroix committed:
[XSG] correctly pass scopes

Pass variables scopes along, to allow datatemplates
